File: java/base/src/main/java/vsdk/toolkit/io/context/ObjectScopeContext.java
package vsdk.toolkit.io.context;

import java.util.Arrays;

public class ObjectScopeContext {
    public String[] objectNamesList;
    public int objectMaxName;
    public int objectNames;

    private static final int OBJECT_NAMES_ALLOC_INCREMENT = 16;

    public ObjectScopeContext() {
        objectNamesList = null;
        objectMaxName = 0;
        objectNames = 0;
    }

    public void destroy() {
        clear();
    }

    public int pushName(String name) {
        if (name == null) {
            return ParseErrorContext.MGF_ERROR_ILLEGAL_ARGUMENT_VALUE;
        }

        try {
            if (objectNames >= objectMaxName - 1) {
                if (objectMaxName == 0) {
                    objectMaxName = OBJECT_NAMES_ALLOC_INCREMENT;
                    objectNamesList = new String[objectMaxName];
                }
                else {
                    final int previousMaxName = objectMaxName;
                    objectMaxName += OBJECT_NAMES_ALLOC_INCREMENT;
                    objectNamesList = Arrays.copyOf(objectNamesList, objectMaxName);
                    if (objectNamesList == null) {
                        objectMaxName = previousMaxName;
                    }
                }
                if (objectNamesList == null) {
                    objectMaxName = 0;
                    objectNames = 0;
                    return ParseErrorContext.MGF_ERROR_OUT_OF_MEMORY;
                }
            }

            objectNamesList[objectNames] = name;
            objectNames++;
            objectNamesList[objectNames] = null;
            return ParseErrorContext.MGF_OK;
        }
        catch (OutOfMemoryError outOfMemoryError) {
            objectMaxName = 0;
            objectNames = 0;
            objectNamesList = null;
            return ParseErrorContext.MGF_ERROR_OUT_OF_MEMORY;
        }
    }

    public int popName() {
        if (objectNames < 1) {
            return ParseErrorContext.MGF_ERROR_UNMATCHED_CONTEXT_CLOSE;
        }
        objectNames--;
        objectNamesList[objectNames] = null;
        return ParseErrorContext.MGF_OK;
    }

    public void clear() {
        if (objectNamesList != null) {
            for (int i = 0; i < objectNames; i++) {
                objectNamesList[i] = null;
            }
        }
        objectNamesList = null;
        objectMaxName = 0;
        objectNames = 0;
    }
}

File: typescript/base/src/main/vsdk/toolkit/io/context/ObjectScopeContext.ts
import { ParseErrorContext } from "./ParseErrorContext";

export class ObjectScopeContext {
  public objectNamesList: Array<string | null> | null;
  public objectMaxName: number;
  public objectNames: number;

  private static readonly OBJECT_NAMES_ALLOC_INCREMENT = 16;

  public constructor() {
    this.objectNamesList = null;
    this.objectMaxName = 0;
    this.objectNames = 0;
  }

  public destroy(): void {
    this.clear();
  }

  public pushName(name: string | null): number {
    if (name === null) {
      return ParseErrorContext.MGF_ERROR_ILLEGAL_ARGUMENT_VALUE;
    }

    try {
      if (this.objectNames >= this.objectMaxName - 1) {
        if (this.objectMaxName === 0) {
          this.objectMaxName = ObjectScopeContext.OBJECT_NAMES_ALLOC_INCREMENT;
          this.objectNamesList = new Array<string | null>(this.objectMaxName).fill(null);
        }
        else {
          const previousMaxName = this.objectMaxName;
          this.objectMaxName += ObjectScopeContext.OBJECT_NAMES_ALLOC_INCREMENT;
          const newNames = new Array<string | null>(this.objectMaxName).fill(null);
          for (let i = 0; this.objectNamesList !== null && i < this.objectNamesList.length && i < newNames.length; i++) {
            newNames[i] = this.objectNamesList[i] ?? null;
          }
          this.objectNamesList = newNames;
          if (this.objectNamesList === null) {
            this.objectMaxName = previousMaxName;
          }
        }
        if (this.objectNamesList === null) {
          this.objectMaxName = 0;
          this.objectNames = 0;
          return ParseErrorContext.MGF_ERROR_OUT_OF_MEMORY;
        }
      }

      (this.objectNamesList as Array<string | null>)[this.objectNames] = name;
      this.objectNames++;
      if (this.objectNames < (this.objectNamesList as Array<string | null>).length) {
        (this.objectNamesList as Array<string | null>)[this.objectNames] = null;
      }
      return ParseErrorContext.MGF_OK;
    }
    catch (_error) {
      this.objectMaxName = 0;
      this.objectNames = 0;
      this.objectNamesList = null;
      return ParseErrorContext.MGF_ERROR_OUT_OF_MEMORY;
    }
  }

  public popName(): number {
    if (this.objectNames < 1) {
      return ParseErrorContext.MGF_ERROR_UNMATCHED_CONTEXT_CLOSE;
    }
    this.objectNames--;
    if (this.objectNamesList !== null) {
      this.objectNamesList[this.objectNames] = null;
    }
    return ParseErrorContext.MGF_OK;
  }

  public clear(): void {
    if (this.objectNamesList !== null) {
      for (let i = 0; i < this.objectNames; i++) {
        this.objectNamesList[i] = null;
      }
    }
    this.objectNamesList = null;
    this.objectMaxName = 0;
    this.objectNames = 0;
  }
}
